plutus-tx/src/Language/PlutusTx/Lift.hs

```diff
@@ -16,90 +16,111 @@ module Language.PlutusTx.Lift (
+type Throwable uni = (PLC.GShow uni, PLC.GEq uni, PLC.Closed uni, uni `PLC.Everywhere` PrettyConst, GHC.Typeable uni)
 
 -- | Get a Plutus Core term corresponding to the given value.
 safeLift
-    :: (Lift.Lift uni a
-       , PIR.AsTypeError e (PIR.Term TyName Name uni ()) uni (Provenance ()), PLC.GShow uni, PLC.GEq uni
+    :: (Lift.Lift uni fun a
+       , PIR.AsTypeError e (PIR.Term TyName Name uni fun ()) uni fun (Provenance ()), PLC.GShow uni, PLC.GEq uni
        , PIR.AsTypeErrorExt e uni (Provenance ())
-       , PLC.DefaultUni PLC.<: uni
-       , AsError e uni (Provenance ()), MonadError e m, MonadQuote m
+
+       , AsError e uni fun (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.ToBuiltinMeaning uni fun
        )
-    => a -> m (UPLC.Term Name uni ())
+    => a -> m (UPLC.Term Name uni fun ())
 safeLift x = do
     lifted <- liftQuote $ runDefT () $ Lift.lift x
-    -- note: we typecheck&compile the plutus-tx term inside an empty builtin context (PLC.defConfig)
-    compiled <- flip runReaderT defaultCompilationCtx $ compileTerm True lifted
+    tcConfig <- PLC.getDefTypeCheckConfig $ Original ()
+    compiled <- flip runReaderT (toDefaultCompilationCtx tcConfig) $ compileTerm True lifted
     pure $ void $ UPLC.erase compiled
 
 -- | Get a Plutus Core program corresponding to the given value.
 safeLiftProgram
-    :: (Lift.Lift uni a
-       , PIR.AsTypeError e (PIR.Term TyName Name uni ()) uni (Provenance ()), PLC.GShow uni, PLC.GEq uni
+    :: (Lift.Lift uni fun a
+       , PIR.AsTypeError e (PIR.Term TyName Name uni fun ()) uni fun (Provenance ()), PLC.GShow uni, PLC.GEq uni
        , PIR.AsTypeErrorExt e uni (Provenance ())
-       , PLC.DefaultUni PLC.<: uni
-       , AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
-    => a -> m (UPLC.Program Name uni ())
+
+       , AsError e uni fun (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.ToBuiltinMeaning uni fun
+       )
+    => a -> m (UPLC.Program Name uni fun ())
 safeLiftProgram x = UPLC.Program () (PLC.defaultVersion ()) <$> safeLift x
 
 safeLiftCode
-    :: (Lift.Lift uni a
-       , PIR.AsTypeError e (PIR.Term TyName Name uni ()) uni (Provenance ()), PLC.GShow uni, PLC.GEq uni       , PLC.DefaultUni PLC.<: uni
+    :: (Lift.Lift uni fun a
+       , PIR.AsTypeError e (PIR.Term TyName Name uni fun ()) uni fun (Provenance ()), PLC.GShow uni, PLC.GEq uni
        , PIR.AsTypeErrorExt e uni (Provenance ())
-       , AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
-    => a -> m (CompiledCode uni a)
+       , AsError e uni fun (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.ToBuiltinMeaning uni fun
+       )
+    => a -> m (CompiledCode uni fun a)
 safeLiftCode x = DeserializedCode <$> safeLiftProgram x <*> pure Nothing
 
-unsafely :: Throwable uni => ExceptT (Error uni (Provenance ())) Quote a -> a
+unsafely
+    :: (Throwable uni, GHC.Typeable fun, Pretty fun)
```

Comment:
@michaelpj Perhaps we should turn `Throwable uni` into `Throwable uni fun` and add the two new constraints there?

Reply:
Done.